### PR TITLE
enhancement/14010-new-demo-technical-indicators

### PR DESCRIPTION
**Pull request by karolkolodziej**
Feature #14010, added the demo with all technical indicators.

Edited demo: http://jsfiddle.net/BlackLabel/czpj5h74/

### CODE REVIEW

File: samples/stock/demo/all-indicators/demo.html [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+<script src="https://code.highcharts.com/stock/modules/data.js"></script>
+<script src="https://code.highcharts.com/stock/modules/exporting.js"></script>
+<script src="https://code.highcharts.com/stock/indicators/indicators-all.js"></script>
+
+<div class='main'>
+  <div class='selectors'>
+    <div class="left">
+      <label class='left__labels' for="overlays">Overlays:</label>
+      <select class='left__select' id="overlays">
+        <option value="none">None</option>
```
Comment (raf18seb) [inline samples/stock/demo/all-indicators/demo.html:11]: I haven't seen many HTML tags in our official demos. We could consider using Highcharts tools e.g. `H.createElement` and then we could position 'selects' using Highcharts coordinates (without CSS). @pawelfus what do you think?

File: samples/stock/demo/all-indicators/demo.js [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        }, {
+            type: 'column',
+            id: 'volume',
+            name: 'Volume',
+            data: volume,
+            yAxis: 1
+        }]
+    });
+
+    document.getElementById("overlays").addEventListener("change", function (e) {
```
Comment (raf18seb) [inline samples/stock/demo/all-indicators/demo.js:64]: I would put this code either into the chart's callback or load event.

File: samples/stock/demo/all-indicators/demo.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        const seriesToAdd = e.target.value;
+
+        // Destroy previously existing series.
+        if (chart.series[2] && chart.series[2].yAxis.options.index === 0) {
+            chart.series[2].destroy(false);
+        }
+        if (chart.series[3] && chart.series[3].yAxis.options.index === 0) {
+            chart.series[3].destroy(false);
+        }
```
Comment (raf18seb) [inline samples/stock/demo/all-indicators/demo.js:73]: What is the purpose of this if statement?

File: samples/stock/demo/all-indicators/demo.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+            }
+
+            if (seriesToAdd !== 'aroon') {
+                chart.addSeries({
+                    type: seriesToAdd,
+                    linkedTo: 'aapl',
+                    yAxis: 3,
+                    showInLegend: true
+                }, false);
+            }
```
Comment (raf18seb) [inline samples/stock/demo/all-indicators/demo.js:189]: We could put it into one `if-else` statement.

**Review by raf18seb — CHANGES_REQUESTED** [review]

File: samples/stock/demo/all-indicators/demo.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
+Highcharts.getJSON('https://demo-live-data.highcharts.com/aapl-ohlcv.json', function (data) {
+    // split the data set into ohlc and volume
+    var ohlc = [],
+        volume = [],
+        dataLength = data.length,
+
+        i = 0;
+
+    for (i; i < dataLength; i += 1) {
```
Comment (raf18seb) [inline samples/stock/demo/all-indicators/demo.js:9]: We can shorten this to one line.

**Review by raf18seb — CHANGES_REQUESTED** [review]

File: samples/stock/demo/all-indicators/demo.html (hunk 1 above)
Comment (pawelfus) [reply samples/stock/demo/all-indicators/demo.html:11]: I think simple HTML in the demo is fine. For example Periodic table: http://utils.highcharts.local/samples/#view/highcharts/series-heatmap/periodic-table

File: samples/stock/demo/all-indicators/demo.js [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+    // create the chart
+    const chart = Highcharts.stockChart('container', {
+        chart: {
+            height: 600
+        },
+        title: {
+            text: 'AAPL Historical'
+        },
+        subTitle: {
```
Comment (pawelfus) [inline samples/stock/demo/all-indicators/demo.js:32]: `subtitle`

File: samples/stock/demo/all-indicators/demo.js [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    });
+
+    document.getElementById("oscilators").addEventListener("change", function (e) {
+        const seriesToAdd = e.target.value;
+
+        // Add third axis if needed.
+        if (chart.yAxis.length === 3) {
+            chart.yAxis[0].update({
+                height: '55%'
+            });
```
Comment (pawelfus) [inline samples/stock/demo/all-indicators/demo.js:131]: When using multiple updates, set `redraw=false` (the same applies to the `update()` and `addAxis()` below)

File: samples/stock/demo/all-indicators/demo.js [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+                        styles: {
+                            lineColor: 'red'
+                        }
+                    },
+                    senkouSpan: {
+                        color: 'rgba(0, 255, 0, 0.3)',
+                        styles: {
+                            fill: 'rgba(0, 0, 255, 0.1)'
+                        }
+                    }
```
Comment (pawelfus) [inline samples/stock/demo/all-indicators/demo.js:111]: Do we need to define all styles for lines in IKH?

File: samples/stock/demo/all-indicators/demo.js [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+            text: 'All indicators'
+        },
+        legend: {
+            enabled: true
+        },
+        rangeSelector: {
+            selected: 2
+        },
+        tooltip: {
+            split: true
```
Comment (pawelfus) [inline samples/stock/demo/all-indicators/demo.js:42]: This is the default, isn't it?

File: samples/stock/demo/all-indicators/demo.css [hunk 10]
```diff
@@ -0,0 +1,42 @@
+.main {
+	position: absolute;
+	min-width: 800px;
+}
+
+.selectors {
+	height: 20px;
+	font-size: 20px;
+	display: flex;
```
[the remark this answers is not in the document]
Comment (pawelfus) [reply samples/stock/demo/all-indicators/demo.css:9]: + `flex` is tricky when comes to auto-resizing the chart

File: samples/stock/demo/all-indicators/demo.js (hunk 8 above)
Comment (karolkolodziej) [reply samples/stock/demo/all-indicators/demo.js:111]: No, we don't have to, actually most of them (especially the fill) are set by default.

File: samples/stock/demo/all-indicators/demo.js (hunk 3 above)
Comment (karolkolodziej) [reply samples/stock/demo/all-indicators/demo.js:73]: User can add first an oscillator (which will be series with the index 2), then can add an overlay (which will be with index 3). The only thing (as far as I can see) which you can use to distinguish them is the yAxis index.
The main purpose for that statements is to detect which series is assigned to which axis and destroy the correct one.
Is there any room for improvement?

File: samples/stock/demo/all-indicators/demo.css (hunk 10 above)
[the remark this answers is not in the document]
Comment (karolkolodziej) [reply samples/stock/demo/all-indicators/demo.css:9]: I removed the `flex` property and I used some different style attributes.

**Review by raf18seb — APPROVED** [review]
Looks good, thanks!

File: samples/stock/demo/all-indicators/demo.js [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+            } else {
+                if (seriesToAdd === 'aroon') {
+                    chart.addSeries({
+                        yAxis: 3,
+                        type: 'aroon',
+                        linkedTo: 'aapl',
+                        color: 'green',
+                        lineWidth: 1,
+                        showInLegend: true,
+                        aroonDown: {
```
Comment (pawelfus) [inline samples/stock/demo/all-indicators/demo.js:129]: Can we simplify it? I mean, let's treat `aroon` as any other tech. indicator, without defining styles (just use defaults)

File: samples/stock/demo/all-indicators/demo.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+                chart.addSeries({
+                    type: seriesToAdd,
+                    linkedTo: 'aapl',
+                    showInLegend: true
+                }, false);
+            }
+            chart.redraw();
+        });
+
+        document.getElementById("oscilators").addEventListener("change", function (e) {
```
Comment (pawelfus) [inline samples/stock/demo/all-indicators/demo.js:79]: Sorry for missing this earlier, just a typo: "oscillators"

**Review by pawelfus — COMMENTED** [review]
Can we change the demo, I mean the way it works? For example:

1. Let's init chart and select by default one overlay and one oscillator (with its own yAxis)
2. Remove `None` option from dropdowns. That simplifies a lot demo (managing yAxis heights, indexes, etc)
3. Add `id` to overlay and oscillator series. So when changing indicator, we simply use `let series = chart.get('series-id'); if (series) { series.remove(false) }` then add new series

**Comment by karolkolodziej** [comment]
I've changed that logic and simplified the demo as you mentioned.

The demo looked strange in utils because of the CSS, I also fixed that.
Updated demo: http://jsfiddle.net/BlackLabel/0qnjfxtv/

**Review by sebastianbochan — CHANGES_REQUESTED** [review]
- change class names
- replace px with em (font-size)
- remove relative / top and use margins / display-inline etc.

File: samples/stock/demo/all-indicators/demo.js [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        ]);
+
+        volume.push([
+            data[i][0], // the date
+            data[i][5] // the volume
+        ]);
+    }
+
+    // create the chart
+    const chart = Highcharts.stockChart('container', {
```
Comment (raf18seb) [inline samples/stock/demo/all-indicators/demo.js:23]: `const` will not work in the old browsers. We can remove it (see the comment below).

File: samples/stock/demo/all-indicators/demo.js [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+            showInLegend: true,
+            yAxis: 0
+        }, {
+            type: 'macd',
+            id: 'oscillator',
+            linkedTo: 'aapl',
+            showInLegend: true,
+            yAxis: 2
+        }]
+    }, function () {
```
Comment (raf18seb) [inline samples/stock/demo/all-indicators/demo.js:72]: Instead of declaring a `const chart` variable, we can use `chart` as function's argument like
```js
}, function (chart) {
```

File: samples/stock/demo/all-indicators/demo.js [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+        document.getElementById("overlays").addEventListener("change", function (e) {
+            const series = chart.get('overlay');
```
Comment (raf18seb) [inline samples/stock/demo/all-indicators/demo.js:74]: Another `const`. Please, find all of them. The same with double quotes.

Comment (raf18seb) [inline samples/stock/demo/all-indicators/demo.js:74]: Mixing single and double quotes. Prefer single `"` -> `'`

File: samples/stock/demo/all-indicators/demo.js [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+            if (series) {
+                series.remove(false);
+                chart.addSeries({
+                    type: e.target.value,
+                    linkedTo: 'aapl',
+                    id: 'overlay',
+                    yAxis: 0,
```
Comment (raf18seb) [inline samples/stock/demo/all-indicators/demo.js:82]: Is it necessary? `yAxis: 0` is by default. Or maybe this indicator doesn't stick to the first axis by default?

File: samples/stock/demo/all-indicators/demo.js [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+                    showInLegend: true
```
Comment (raf18seb) [inline samples/stock/demo/all-indicators/demo.js:83]: There are many `showInLegend: true`. If all series should be shown in the legend, you could define it in `plotOptions.series`.

File: samples/stock/demo/all-indicators/demo.details [hunk 18]
```diff
@@ -0,0 +1,10 @@
+---
+name: All technical indicators
+authors:
+  - Karol Kołodziej
```
Comment (raf18seb) [inline samples/stock/demo/all-indicators/demo.details:4]: By the occasion - don't you want to replace `ł` with `l` as we discussed?

File: samples/stock/demo/all-indicators/demo.html [hunk 19]
```diff
@@ -0,0 +1,64 @@
+<script src="https://code.highcharts.com/stock/highstock.js"></script>
+<script src="https://code.highcharts.com/stock/modules/data.js"></script>
+<script src="https://code.highcharts.com/stock/modules/exporting.js"></script>
+<script src="https://code.highcharts.com/stock/indicators/indicators-all.js"></script>
+
+<div class="main-wrapper">
+  <div class="selectors-container">
```
Comment (raf18seb) [inline samples/stock/demo/all-indicators/demo.html:7]: You can leave it as it is, but do we need this `<div class="selectors-container">`? I don't see any good reasons (maybe there is one) why this
```html
<div class="a">
    <div class="b">
        <p></p>
        <p></p>
        <p></p>
    </div>
</div>
```
cannot be replaced with (with corrected CSS styles and classes maybe)
```html
<div class="a b">
    <p></p>
    <p></p>
    <p></p>
</div>
```

Just a suggestion for the future.

File: samples/stock/demo/all-indicators/demo.html [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+    <div class="lhs-col">
+      <label for="overlays">Overlays:</label>
+      <select class='left-select' id="overlays">
```
Comment (raf18seb) [inline samples/stock/demo/all-indicators/demo.html:10]: In this case, a double quote is preferable `"`.

**Review by raf18seb — CHANGES_REQUESTED** [review]

File: samples/stock/demo/all-indicators/demo.html (hunk 19 above)
Comment (karolkolodziej) [reply samples/stock/demo/all-indicators/demo.html:7]: `main-wrapper` class is to... well.. wrap the whole selectors and the chart together as it might be a part of some bigger design/site

**Review by raf18seb — APPROVED** [review]
Thanks!

**Review by pawelfus — APPROVED** [review]
Thanks!

File: samples/stock/demo/all-indicators/demo.html [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+        <option value="vwap">WMA (Weighted Moving Average)</option>
+        <option value="wma">VWAP (Volume Weighted Average Price)</option>
+        <option value="zigzag">Zig Zag</option>
+      </select>
+    </div>
+
+    <div class="rhs-col">
+      <label for="oscillators">Oscillators:</label>
+      <select class="right-select" id="oscillators">
+        <option value="none">None</option>
```
Comment (pawelfus) [inline samples/stock/demo/all-indicators/demo.html:35]: Accepted, and then realized: There is no `none` technical indicator :)

File: samples/stock/demo/all-indicators/demo.html [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+        <option value="apo">Absolute price indicator</option>
+        <option value="ad">A/D (Accumulation/Distribution)</option>
+        <option value="aroon">Aroon</option>
+        <option value="aroonoscillator">Aroon oscilator</option>
```
Comment (pawelfus) [inline samples/stock/demo/all-indicators/demo.html:39]: Oscillator :)

File: samples/stock/demo/all-indicators/demo.html [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+        <option value="atr">ATR (Average True Range)</option>
+        <option value="ao">Awesome oscilator</option>
```
Comment (pawelfus) [inline samples/stock/demo/all-indicators/demo.html:41]: Oscillator

**Review by pawelfus — CHANGES_REQUESTED** [review]
minor polishing

**Review by pawelfus — APPROVED** [review]
Thanks!

**Review by pawellysy — APPROVED** [review]
Looks good!

**Review by TorsteinHonsi — APPROVED** [review]
Thanks, nice demo! I did some final CSS polishing, now merging.
